File: modules/RootTipMulti/RootTipMulti.pl
#!/usr/bin/perl
package BQPhytomorph::RootTipMulti;

use warnings;
use strict;

use Cwd 'abs_path';
use File::Basename;
use File::Copy;
use Getopt::Long;
use XML::LibXML;
use XML::Simple;
use Data::Dumper;
use POSIX qw/strftime/;
use URI::Escape;
use Log::Log4perl;

use BQPhytomorph;

our $EXEC          = "./maizeG";
my $log = Log::Log4perl::get_logger('roottipmulti');

#########################################
## Argument handling
my ($help, $image, $mex, $user, $password, $userpass, $staging_path, $token);
my $debug = 0;
my $dryrun = 0;

usage() if ( ! GetOptions('help|?' => \$help,
                          'debug|d' => \$debug,
                          'dryrun|n' => \$dryrun,
                          'mex_url=s' => \$mex,
                          'image_url=s' => \$image,
                          'staging_path=s'=>\$staging_path,
                          'user=s' => \$user,
                          'password=s' => \$password,
                          'token=s'=>\$token,
                         )
          or defined $help );
my $options = { debug=> $debug, dryrun => $dryrun };
my $command  = shift;

usage() if (!defined $command);


if ($user && $password) {
  $userpass = $user . ':' . $password;
}
print "command is " . $command . "\n";

my $ret=0;
$ret=command_setup() if ($command eq 'setup');
$ret=command_teardown() if ($command eq 'teardown');
$ret=command_start() if ($command eq 'start');

exit($ret);


sub usage {
  print "Unknown option: @_\n" if ( @_ );
  print "usage: RootTip [--mex=URL] [--userpass=USER:PASS] [--image_url=IMAGE] [--staging=PATH] [--debug] [--help]  [start|setup|teardown] \n";
  exit 1;
}

sub command_fetch {
  my $bq = new BQPhytomorph(mex => $mex, userpass => $userpass ) if ( $mex) ;
  fetchGObjectsAsCSV($bq, $mex, $staging_path);
}


sub command_setup {
  print "SETUP\n";
  my $bq = new BQPhytomorph(mex_url => $mex, userpass => $userpass,
                            token=> $token, options=>$options );
  $bq->update_mex("initializing");

  $bq->fetchImagePlanes ($image,  $staging_path);
  fetchGObjectsAsCSV($bq, $mex, $staging_path);

  $bq->update_mex('scheduling');
  return 0;
}

sub command_start {
  print "START $EXEC $staging_path/ \n";
  my $bq = new BQPhytomorph(mex_url => $mex, userpass => $userpass,
                            token=> $token, options=>$options );

  $bq->update_mex("running");
  return system ($EXEC,  "$staging_path/");
}


### This command is called by Condor
sub command_teardown {
  print "TEARDOWN\n";

  my $bq = new BQPhytomorph(mex_url => $mex, userpass => $userpass,
                            token=> $token, options=>$options );

  # Collect the condor logs and determine the status

  # Collect results, process and save
  my $tips = './tips.csv';
  my $angles = './angles.csv';
  my $tags;
  if (-f $tips && -f $angles ) {
    eval {
      $tags = saveTipsAndAngles ($bq, $mex, $image, $tips, $angles ) ;
      1;
    } or do {
      $bq->finish_mex ( status=>"FAILED", msg=>$@  );
      return 1;
    }
  }
  # Update the MEX with state
  $bq->finish_mex ( status=>"FINISHED",  tags=>$tags);
  return 0;
}

sub fetchGObjectsAsCSV {
  my ($bq, $mexurl, $staging) = @_;

  # Parse the gobject creating a CSV

  my @tips;
  my $count = 0;

  my $mexdoc = $bq->fetchxml ($mexurl."?view=deep");
  die "fetch failed:" . $mexurl unless defined $mexdoc;

  my @vertices = $mexdoc->findnodes ('//vertex');
  for my $v (@vertices) {
    my $x = int ($v->getAttribute('x'));
    my $y = int ($v->getAttribute('y'));
    push (@tips ,  "$y, $x");
    #print "$y, $x";
  }
  if (scalar @tips == 0) {
      die "Can't get input points from mex" ;
  }

  my $tipout = abs_path ($staging .'/inputtips.csv');

  open(TIPS, ">$tipout");
  foreach my $tip (@tips) { 
    print TIPS "$tip\n";
  }
  close(TIPS);
}


sub saveTipsAndAngles {
  my ($bq, $mex, $image, $tips, $angles) = @_;

  open (TIPS, "<$tips") or die "No $tips";
  my @t = <TIPS>;
  chomp @t;
  close(TIPS);
  open (ANG, "<$angles") or die "No $angles";
  my @a = <ANG>;
  chomp @a;
  close(ANG);
#   open (GROWTH, "<$growth") or die "No $angles";
#   my @g = <GROWTH>;
#   chomp @g;
#   close(GROWTH);

  #print "TIPS".Dumper(@tips);
  #print "ANG".Dumper(@a);

  my $xs = new XML::Simple(RootName => undef);

  my $imagedoc = $bq->fetchxml ($image);
  die "fetch failed:" . $image unless defined $imagedoc;

  my $xsize = $imagedoc->findnodes ('//image/@x')->to_literal->value;
  my $ysize = $imagedoc->findnodes ('//image/@y')->to_literal->value;

  # <gobject name="Tip 1" >
  #    <gobject type=tipangle>
  #      <point x= y=  t=1..N  >
  #      <tag name="angle" value="" />
  #    <gobject>
  #    ...
  # </gobject>
  # <gobject name="Tip 2" >
  #
  my $count = @a;
  my @gobs ;
  my @angles = split (/,/, $a[0]);
  my $tip_count = @angles;
  for (my $i=0; $i<$tip_count ; $i++){
    push @gobs, [];
  }

  for (my $i=0; $i < $count; $i++) {
    #print "$i $a[$i]\n";
    #my ($x,$y) = split(/,/, $t[$i]);
    my @points = split(/,/, $t[$i]);
    my @angles = split(/,/, $a[$i]);
    #my @growths = split(/,/, $g[$i]);

    for (my $g=0; $g<$tip_count; $g++){
      push @{$gobs[$g]}, 
	{ type=>"tipangle",
	  point =>  {   vertex => [ { x=> $points[$g*2], y=> $points[$g*2+1], t=> $i} ]  },
	  tag   => [ { name=>"angle", value=> $angles[$g]},
		     #				      { name=>"growth", value=> $growths[$g]}
		   ]
	};
    }
  }

  my $gentime = strftime ("%Y-%m-%d %H:%M:%S", localtime);
  my $req = {request => {  gobject =>
			   [{ name => "RootTip  $gentime",
			      gobject => [] } ]
                        }
            };

  for (my $g=0; $g<$tip_count; $g++){
    push @{$req->{request}->{gobject}[0]->{gobject}}, { name => "Tip-" . ($g+1),
							gobject=>\@{$gobs[$g]}};
  }


  my $content =  $xs->XMLout($req);
  #print $content;
  my $gurl    = "";
  my $doc;
  if ( ! $dryrun ) {
    $doc = $bq->postxml ($image . "/gobjects", $content);
    die "post failed:" . $image . "/gobjects" unless defined $doc;
    $gurl = $doc->findnodes('//gobject/@uri')->to_literal->value;
  }

  if ( $debug ) {
    print "GOB $gurl\n";
    print "GOB => $content";
  }

  my $xpath="";
  for (my $g=0; $g<$tip_count; $g++){
    if (! $xpath) {
      $xpath = "&xpath=";
    } else { 
      $xpath .= "&xpath" . ($g).'=';
    }
    $xpath .= uri_escape('//gobject[@name="' . "Tip-" . ($g+1) . '"]//tag[@name="angle"]');
  }

  $req = { request => 
           { tag =>  
             [ { name => "RootTip",
                 tag => [ { name=>"gobjects_url", type=>"link", value=>$gurl},
                          { name=>"mex_url", type=>"link", value=>$mex},
                          { name=>'date_time', value=> $gentime },
                          { name=>"area-histogram", type=>"statistics",
                            value => ("http:/stats?url=" . uri_escape($gurl)
                                      .$xpath
                                      ."&xmap=".uri_escape("tag-value-number")
                                      ."&xreduce=vector&run=true")},
                        ]}]}};





  $content =  $xs->XMLout($req);

  if (! $dryrun ) {
    $doc = $bq->postxml ($image . "/tags", $content);
    die "post failed:" . $image . "/tags" unless defined $doc;
  }
  if ( $debug ) {
    print "TAG => $content";
  }

  my $mextags = [ {name=>"gobjects_url", type=>"link", value=>$gurl} ];
  return $mextags;
}

File: modules/RootTipMulti/RootTipMulti.py
#!/usr/bin/env python
import os
import sys
import optparse
import subprocess
import glob
import csv
import pickle
import logging
import itertools

from bqapi import BQSession
from bqapi.util import fetch_image_planes, AttrDict
from lxml.builder import E


logging.basicConfig(level=logging.DEBUG)

EXEC = "./maizeG"

def gettag (el, tagname):
    for kid in el:
        if kid.get ('name') == tagname:
            return kid, kid.get('value')
    return None,None

class RootTip(object):

    def setup(self):
        #if not os.path.exists(self.images):
        #    os.makedirs(self.images)

        self.bq.update_mex('initializing')

        results = fetch_image_planes(self.bq, self.resource_url, '.')

        # extract gobject inputs
        tips = self.bq.mex.find('inputs', 'tag').find('image_url', 'tag').find('tips', 'gobject')
        with open('inputtips.csv', 'w') as TIPS:
            for point in tips.gobjects:
                print >>TIPS, "%(y)s, %(x)s" % dict(x=point.vertices[0].x,y=point.vertices[0].y)



    def start(self):
        self.bq.update_mex('executing')
        # Matlab requires trailing slash
        subprocess.call([EXEC, './'])


    def teardown(self):
        # Post all submex for files and return xml list of results
        gobjects = self._read_results()
        tags = [{ 'name': 'outputs',
                  'tag' : [{'name': 'rootimage', 'type':'image', 'value':self.resource_url,
                            'gobject' : [{ 'name': 'root_tips', 'type': 'root_tips', 'gobject' : gobjects }] }]
                  }]
        self.bq.finish_mex(tags = tags)

    def _read_results(self, ):
        results  = []
        #image = self.bq.load(self.resource_url, view='full')
        #xmax, ymax, zmax, tmax, ch = image.geometry()
        #  each line is a time point, tracked points in line
        tips = [ t for t in csv.reader(open('tips.csv','rb')) ]
        # each line is a time point, tracked points per line pt1, pt2, pt2
        angles = [ a for a in csv.reader(open('angles.csv','rb')) ]
        # We want to generate lines in time to keep the tracked point in a single object
        # <gobject name="Tip 1" >
        #    <gobject type=tipangle>
        #      <point x= y=  t=1..N  >
        #      <tag name="angle" value="" />
        #    <gobject>
        #    ...
        # </gobject>
        # <gobject name="Tip 2" >
        #
        planes = len(angles)
        tip_count   = planes and len(angles[0]) or 0
        tracks = []
        for pt in range(tip_count):
            gobs = []
            for t_plane in range(planes):
                gobs.append(
                    {'type': 'tipangle',
                     'tag': [{ 'name': 'angle', 'value': angles[t_plane][pt] }],
                     'point':[{'vertex' :
                               [{ 't': str(t_plane), 'x': str(tips[t_plane][pt*2]),
                                  'y': str(tips[t_plane][pt*2+1]) }]
                               }]
                     })
            tracks.append({'name': 'tip-%s' % pt, 'type': 'roottip', 'gobject': gobs})

        return tracks



    def run(self):
        parser  = optparse.OptionParser()
        parser.add_option('-d','--debug', action="store_true")
        parser.add_option('-n','--dryrun', action="store_true")
        parser.add_option('--credentials')
        parser.add_option('--image_url')

        (options, args) = parser.parse_args()
        named = AttrDict (bisque_token=None, mex_url=None, staging_path=None, image_url=None)
        for arg in list(args):
            tag, sep, val = arg.partition('=')
            if sep == '=':
                named[tag] = val
                args.remove(arg)
        self.named_args = named
        self.staging_path = named.get('staging_path')

        if named.bisque_token:
            self.bq = BQSession().init_mex(named.mex_url, named.bisque_token)
            self.resource_url = self.bq.parameter_value ('image_url')
        elif options.credentials:
            user,pwd = options.credentials.split(':')
            self.bq = BQSession().init_local(user,pwd)
            self.resource_url =  named.image_url
        else:
            parser.error('need bisque_token or user credential')

        if self.resource_url is None:
            parser.error('Need a resource_url')

        if not args :
            commands = ['setup', 'start', 'teardown']
        else:
            commands = [ args ]

        try:
            for command in commands:
                command = getattr(self, command)
                r = command()
        except Exception, e:
            logging.exception ("problem during %s" % command)
            self.bq.fail_mex(msg = "Exception during %s: %s" % (command,  e))
            sys.exit(1)

        sys.exit(0)

        #command = args.pop(0)

        #if command not in ('setup','teardown', 'start'):
        #    parser.error('Command must be start, setup or teardown')

        # maltab code requires trailing slash..
        #self.images = os.path.join(options.staging_path, 'images') + os.sep
        #self.image_map_name = os.path.join(options.staging_path, IMAGE_MAP)
        #self.resource_url = options.resource_url
        #self.config = options
        #self.is_dataset = 'dataset' in self.resource_url


        #command = getattr(self, command)
        #command()





if __name__ == "__main__":
    RootTip().run()
